### PR TITLE
[MINOR][ML] Increase Bounded MLOR (without regularization) test error tolerance

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?
Improve LogisticRegression test error tolerance

### Why are the changes needed?
When we switch BLAS version, some of the tests will fail due to too strict error tolerance in test.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
N/A

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 1]
```diff
@@ -2962,7 +2962,14 @@ object LogisticRegressionSuite {
   def checkCoefficientsEquivalent(coefficients1: Matrix, coefficients2: Matrix): Unit = {
     coefficients1.colIter.zip(coefficients2.colIter).foreach { case (col1: Vector, col2: Vector) =>
       (col1.asBreeze - col2.asBreeze).toArray.toSeq.sliding(2).foreach {
-        case Seq(v1, v2) => assert(v1 ~= v2 absTol 1E-3)
+        case Seq(v1, v2) =>
+          if (math.abs(v1) < 1E-5 || math.abs(v2) < 1E-5) {
+            // If one of v1 and v2 is very close to zero, then compare absTol
+            assert(v1 ~= v2 absTol 1E-2, s"v1 ~= v2 absTol 0.01 faild: v1=${v1}, v2=${v2}")
+          }
```
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala:2969]: nit： lint

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 2]
```diff
@@ -2962,7 +2962,14 @@ object LogisticRegressionSuite {
...
+          else {
+            assert(v1 ~= v2 relTol 1E-2, s"v1 ~= v2 relTol 0.01 faild: v1=${v1}, v2=${v2}")
```
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala:2971]: I guess you can keep previous logic `v1 ~= v2 absTol 1E-3` here.

**Review by srowen — COMMENTED** [review]
Hm, is this the best fix? a difference of 0.0001 vs 0.0002 could be significant. Do we just need to run LR more iterations instead?

**Comment by WeichenXu123** [comment]
@srowen 

> a difference of 0.0001 vs 0.0002 could be significant.

One possibility is that, the result value will converge to 0, but in limited iterations, it may result in some near zero value like such as 1e-4 or 2e-4, @zhengruifeng Could you help check whether increase iteration will help make than converge fast enough to be zero ? Not sure the best threshold to set here for go into the branch of absTol test.

**Comment by srowen** [comment]
If that's true, then isn't the 'right' expected value 0, and we're looking for a value at most some epsilon?

**Comment by WeichenXu123** [comment]
> If that's true, then isn't the 'right' expected value 0, and we're looking for a value at most some epsilon?

Reasonable. We should set expected value to be 0. Need check test code.

> If that's true, then isn't the 'right' expected value 0, and we're looking for a value at most some epsilon?

Resonable. Found some other issue in test code. Let me check.

**Comment by zhengruifeng** [comment]
@srowen @WeichenXu123 

I had found that solver `BreezeLBFGSB` for Huber seems quite unstable, see [SPARK-32060](https://issues.apache.org/jira/browse/SPARK-32060) for details.

**Comment by WeichenXu123** [comment]
The master code use some werid logic to compare:
```
(col1.asBreeze - col2.asBreeze).toArray.toSeq.sliding(2).foreach {
        case Seq(v1, v2) => assert(v1 ~= v2 absTol 1E-3)
      }
```
change it to compare left and right directly.

**Review by srowen — COMMENTED** [review]
Maybe this is fine as a band-aid but still feels funky. Can we simply loosen the tolerance if we can't figure it out, or run more iterations etc to improve stability in the test? maybe even different initial conditions.

**Comment by zhengruifeng** [comment]
```
  /**
   * When no regularization is applied, the multinomial coefficients lack identifiability
   * because we do not use a pivot class. We can add any constant value to the coefficients
   * and get the same likelihood. If fitting under bound constrained optimization, we don't
   * choose the mean centered coefficients like what we do for unbound problems, since they
   * may out of the bounds. We use this function to check whether two coefficients are equivalent.
   */
  def checkCoefficientsEquivalent(coefficients1: Matrix, coefficients2: Matrix): Unit = {
    coefficients1.colIter.zip(coefficients2.colIter).foreach { case (col1: Vector, col2: Vector) =>
      (col1.asBreeze - col2.asBreeze).toArray.toSeq.sliding(2).foreach {
        case Seq(v1, v2) => assert(v1 ~= v2 absTol 1E-3)
      }
    }
  }
```

According to the comments, it looks that this compare logic is designed on purpose. @WeichenXu123

**Comment by WeichenXu123** [comment]
@srowen Is this change ok now?
We need backport this to spark 3.1 branch, the adaptive blockify feature added in spark 3.1 increase the probability of this Bounded MLOR (without regularization) test failure.

**Comment by srowen** [comment]
I think it's OK as a band-aid and OK to back port

**Comment by srowen** [comment]
@WeichenXu123 go ahead and merge and backport if you need it in 3.1

**Comment by WeichenXu123** [comment]
merge to master and branch-3.1, thanks!

**Pull request closed without merging**